plugin/src/main/scala/org/scalamacros/paradise/typechecker/Expanders.scala

```diff
@@ -149,7 +149,9 @@ trait Expanders {
             assert(!sym.rawInfo.isInstanceOf[Namer#MaybeExpandeeCompleter])
           }
           val derivedTrees = attachedExpansion(sym).getOrElse(List(stat))
-          val (me, others) = derivedTrees.partition(_.symbol == sym)
+          // Rewrap DocDef in cases where the same class or object, by name, is output by the macro. This can't just check
+          // s.symbol == sym because in the case of nested classes, the original sym is destroyed at Namers.scala:383.
+          val (me, others) = derivedTrees.partition(s => s.symbol.isClass == sym.isClass && s.symbol.fullName == sym.fullName)
```

Comment:
Nice catch! I can imagine how much you had to debug the problem. Some comments:

1) I think that `Rewrap DocDef in cases where the same class or object, by name, is output by the macro` is redundant here, because rewrapping is done not on this line, but later on.

2) I'd say "`sym` is unlinked immediately after expansion, and a new symbol is created in its stead".

3) The check can be made more robust. How about `s.symbol.owner == sym.owner && s.symbol.name == sym.name`?